source/agora/common/BlockStorage.d

```diff
@@ -0,0 +1,292 @@
+/*******************************************************************************
+
+    Defines the memory mapped file of a block.
+
+    The number of files is one.
+    We will expand to use multiple files in the next step.
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.BlockStorage;
+
+import agora.common.Block;
+import agora.common.Deserializer;
+import agora.common.Serializer;
+
+import std.algorithm.comparison;
+import std.file;
+import std.format;
+import std.mmfile;
+import std.path;
+
+
+/// Ditto
+public class BlockStorage
+{
+    /// Instance of memory mapped file
+    private MmFile file;
+
+    /// Is already mapped ?
+    private bool is_mapped;
+
+    /// Path of block data
+    private string path;
+
+    /// Index is block height, Value is position of file
+    /// We will change to index file in the next step.
+    private size_t[] block_positions;
+
+    /// Ctor
+    public this (string path)
+    {
+        this.is_mapped = false;
+        this.path = path;
+    }
+
+    public ~this ()
+    {
+        this.unMap();
+    }
+
+    /***************************************************************************
+
+        Make a file name using the index.
+
+        Returns:
+            Returns the file name.
+
+    ***************************************************************************/
+
+    private string getFileName ()
+    {
+        return buildPath(this.path, "block.dat");
+    }
+
+    /***************************************************************************
+
+        Open memory mapped file.
+
+        If a size change occurs, open it again.
+
+        Params:
+            resize = The size of the file to change. If 0 does not change it.
+
+    ***************************************************************************/
+
+    private void map (size_t resize = 0)
+    {
+        if (this.is_mapped && (resize == 0))
+            return;
+
+        if (this.is_mapped)
+            this.unMap();
+
+        const string name = this.getFileName();
+
+        if (name.exists())
+            this.file =
+                new MmFile(
+                    name,
+                    MmFile.Mode.readWrite,
+                    resize,
+                    null
+                );
+        else
+            this.file =
+                new MmFile(
+                    name,
+                    MmFile.Mode.readWriteNew,
```

Comment:
It seems that the `readWrite` mode already creates the file if it doesn't exist? From the documentation:

```
/**
 * The mode the memory mapped file is opened with.
 */
enum Mode
{
    read,            /// Read existing file
    readWriteNew,    /// Delete existing file, write new file
    readWrite,       /// Read/Write existing file, create if not existing
    readCopyOnWrite, /// Read/Write existing file, copy on write
}
```

Unless I'm wrong, I think this part can just be:

```diff
diff --git a/source/agora/common/BlockStorage.d b/source/agora/common/BlockStorage.d
index f8ad8a0..a0a1456 100644
--- a/source/agora/common/BlockStorage.d
+++ b/source/agora/common/BlockStorage.d
@@ -90,22 +90,13 @@ public class BlockStorage

         const string name = this.getFileName();

-        if (name.exists())
-            this.file =
-                new MmFile(
-                    name,
-                    MmFile.Mode.readWrite,
-                    resize,
-                    null
-                );
-        else
-            this.file =
-                new MmFile(
-                    name,
-                    MmFile.Mode.readWriteNew,
-                    max(1, resize),
-                    null
-                );
+        this.file =
+            new MmFile(
+                name,
+                MmFile.Mode.readWrite,
+                resize,
+                null
+            );

         this.is_mapped = true;
     }
```

Reply:
I just checked. I'll fix it. Thank you.